lib/que/worker.rb

```diff
@@ -1,106 +1,132 @@
 # frozen_string_literal: true
 
+require "pg"
 require "benchmark"
-require "que/metrics"
-require "que/locker"
-require "que/job_timeout_error"
+require "prometheus/client"
+
+require_relative "locker"
+require_relative "job_timeout_error"
 
 module Que
   class Worker
     # Defines the time a worker will wait before checking Postgres for its next job
-    DEFAULT_QUEUE = ''
+    DEFAULT_QUEUE = ""
     DEFAULT_WAKE_INTERVAL = 5
     DEFAULT_LOCK_CURSOR_EXPIRY = 0 # seconds
-    DEFAULT_STOP_TIMEOUT = 5 # seconds
 
-    # Start worker_count number of threaded workers, returning a proc that can be called
-    # to stop the workers from working.
-    def self.start_workers(worker_count, metrics_labels: {}, **kwargs)
-      Que.logger.info(msg: 'Starting workers', event: 'que.worker.start', worker_count: worker_count)
+    RunningSecondsTotal = Prometheus::Client.registry.counter(
+      :que_worker_running_seconds_total, "Time since starting to work jobs",
+    )
+    SleepingSecondsTotal = Prometheus::Client.registry.counter(
+      :que_worker_sleeping_seconds_total, "Time spent sleeping due to no jobs",
+    )
+    JobsWorkedTotal = Prometheus::Client.registry.counter(
+      :que_job_worked_total, "Counter for all jobs processed",
+    )
+    JobsErrorTotal = Prometheus::Client.registry.counter(
+      :que_job_error_total, "Counter for all jobs that were run but errored",
+    )
+    JobsWorkedSecondsTotal = Prometheus::Client.registry.counter(
+      :que_job_worked_seconds_total, "Sum of the time spent processing each job class",
+    )
+    JobsLatencySecondsTotal = Prometheus::Client.registry.counter(
+      :que_job_latency_seconds_total, "Sum of time spent by job and priority waiting in queue",
+    )
 
-      # Create each worker with a distinct ID in order to differentiate each workers
-      # prometheus metrics.
-      workers = Array.new(worker_count) do |id|
-        Worker.new(**kwargs.merge(metrics_labels: metrics_labels.merge(worker: id)))
+    # Metrics from the worker are often collected over long spans of time. We don't want
+    # to update the metric only once the long-running task has complete, as this leads to
+    # graphs that dump all the weight of your task just as it finishes. The Collector
+    # class is used to provide an interface to update each metric as we receive requests
+    # to read the metric, solving this problem.
+    class Collector
```

Comment:
As far as I can tell, this class only makes sense to use with metrics of the form "time since X". It would be good to make that clear here, to prevent it being misused and also to help explain why it is necessary.

It would also be good to put it in its own file - this one is big enough as it is.